irohad/torii/command_client.cpp

```diff
@@ -12,42 +12,120 @@ limitations under the License.
   using iroha::protocol::ToriiResponse;
 
-  ToriiResponse sendTransaction(const Transaction& tx,
-                                const std::string& targetPeerIp) {
-    CommandClient client(targetPeerIp, 50051);  // TODO: Get port from config
-    return client.Torii(tx);
-  }
+  /*
+   * avoids from multiple-definition of ThreadPool
+   * tp::ThradPool is type alias, not class. So we can't use struct ThreadPool;
+   * We shouldn't know about ThreadPoolImpl, that is template class.
+   */
+  struct ThreadContainer {
+    tp::ThreadPool pool;
+  };
+
+  struct ToriiAsyncClientCall {
+    iroha::protocol::ToriiResponse response;
+    grpc::ClientContext context;
+    grpc::Status status;
+    std::unique_ptr<grpc::ClientAsyncResponseReader<iroha::protocol::ToriiResponse>> response_reader;
+  };
 
-  CommandClient::CommandClient(const std::string& ip, int port) {
-    // TODO(motxx): call validation of ip format and port.
-    auto channel = grpc::CreateChannel(ip + ":" + std::to_string(port),
-                                       grpc::InsecureChannelCredentials());
-    stub_ = iroha::protocol::CommandService::NewStub(channel);
+  CommandClient::CommandClient(const std::string& ip, const int port)
+    : stub_(iroha::protocol::CommandService::NewStub(
+    grpc::CreateChannel(ip + ":" + std::to_string(port), grpc::InsecureChannelCredentials()))),
+      listenerPool_(new ThreadContainer)
+  {}
+
+  CommandClient::~CommandClient() {
+    delete listenerPool_;
   }
 
-  ToriiResponse CommandClient::Torii(const Transaction& tx) {
+  /**
+   * requests tx to a torii server and returns response (blocking, sync)
+   * @param tx
+   * @return ToriiResponse
+   */
+  ToriiResponse CommandClient::ToriiBlocking(const Transaction& tx) {
     ToriiResponse response;
-    auto status = stub_->Torii(&context_, tx, &response);
 
-    if (status.ok()) {
-      return response;
-    } else {
-      response.Clear();
-      response.set_code(iroha::protocol::FAIL);
-      response.set_message("connection failed. cannot send transaction.");
+    std::unique_ptr<grpc::ClientAsyncResponseReader<iroha::protocol::ToriiResponse>> rpc(
+      stub_->AsyncTorii(&context_, tx, &cq_)
+    );
+
+    using State = network::UntypedCall<torii::CommandServiceHandler>::State;
+
+    rpc->Finish(&response, &status_, (void *)static_cast<int>(State::ResponseSent));
+
+    void* got_tag;
+    bool ok = false;
+
+    if (!cq_.Next(&got_tag, &ok)) {  // CompletionQueue::Next() is blocking.
+      throw std::runtime_error("CompletionQueue::Next() returns error");
+    }
+
+    assert(got_tag == (void *)static_cast<int>(State::ResponseSent));
+    assert(ok);
+
+    if (status_.ok()) {
       return response;
     }
+
+    response.set_code(iroha::protocol::ResponseCode::FAIL);
+    response.set_message("RPC failed");
+    return response;
+  }
+
+
+  /*
+   * TODO(motxx): We can't use CommandClient::ToriiNonBlocking() for now. gRPC causes the error
```

Comment:
Currently, no. But maybe not so difficult problem.
We can use `ToriiBlocking()` instead and if solved this problem, we can replace `ToriiNonBlocking()` not so hard.

`ToriiBlocking()`
```
auto response = ToriiBlocking(tx);

ToriiNonBlocking(tx, [](ToriiResponse resp) {
   // callback here
})
```